### PR TITLE
Add python pandoc-xnos

### PR DESCRIPTION
**Pull request by ppenguin**
###### Description of changes

`pandoc-xnos` is a python-based `pandoc` filter (consisting of `pandoc-xnos` which can be called by separate binaries for cross-references to equations, tables, sections, figures) when converting from markdown with pandoc.

See upstream: https://github.com/tomduck/pandoc-xnos

The binary installed can be called with `pandoc --filter pandoc-xnos`.

To use only a specific filter (subset of `xnos`) install the related packages `pandoc-{eq|fig|table|sec}nos` also included in this PR.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/pandoc-eqnos/default.nix [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+  src = fetchFromGitHub {
+    owner = "tomduck";
+    repo = pname;
+    rev = version;
+    sha256 = "sha256-7GQdfGHhtQs6LZK+ZyMmcPSkoFfBWmATTMejMiFcS7Y=";
+  };
+
+  propagatedBuildInputs = [ pandoc-xnos ];
+
```
Comment (fabaff) [inline pkgs/development/python-modules/pandoc-eqnos/default.nix:20]: The upstream repo contains [tests](https://github.com/tomduck/pandoc-eqnos/tree/master/test) but it looks like that they need to be generated on the fly.

```suggestion

  # Different pandoc executables are not available
  doCheck = false;

  pythonImportsCheck = [
    "pandoc_eqnos"
  ];
```

File: pkgs/development/python-modules/pandoc-eqnos/default.nix [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  meta = with lib; {
+    description = "Pandoc-eqnos is a pandoc filter for numbering equations and their references when converting from markdown to other formats.";
```
Comment (fabaff) [inline pkgs/development/python-modules/pandoc-eqnos/default.nix:22]:
```suggestion
    description = "Pandoc filter for numbering equations and their references";
```

see https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes

File: pkgs/development/python-modules/pandoc-eqnos/default.nix [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    homepage = "https://github.com/tomduck/pandoc-eqnos";
+    license = licenses.gpl3;
```
Comment (fabaff) [inline pkgs/development/python-modules/pandoc-eqnos/default.nix:24]: See https://github.com/tomduck/pandoc-eqnos/blob/master/setup.py#L6-L8

```suggestion
    license = licenses.gpl3Only;
```

**Review by fabaff — COMMENTED** [review]
Could you please create a commit per module with `python3Packages.xxx: init at x.x.x`?

The comments are also valid for the other modules.

**Comment by ppenguin** [comment]
> Could you please create a commit per module with `python3Packages.xxx: init at x.x.x`?
> 
> The comments are also valid for the other modules.

done, thanks for the guidance.

**Comment by loicreynier** [comment]
> @fabaff Oops, it was spotted in #164552 that these filters should probably be packaged as `pythonApplication` (not `pythonPackage`), since they are not intended to use via `import`. I'll make the necessary changes (hopefully this week, quite busy now) and change the respective commits and PR to that effect.

I agree that individual filters should be packaged as `pythonApplication` but `pandoc-xnos` is also a library and may be used as a package in some projects. In that regards, should not it be build as  a `pythonPackage` and as `pythonApplication`?

**Comment by ppenguin** [comment]
> 



> > @fabaff Oops, it was spotted in #164552 that these filters should probably be packaged as `pythonApplication` (not `pythonPackage`), since they are not intended to use via `import`. I'll make the necessary changes (hopefully this week, quite busy now) and change the respective commits and PR to that effect.
> 
> I agree that individual filters should be packaged as `pythonApplication` but `pandoc-xnos` is also a library and may be used as a package in some projects. In that regards, should not it be build as a `pythonPackage`?

Also a good point... However the only ones that actually seem to use it as a library are the packages that are part of the `pandoc-xnos-suite` (i.e. `pandoc-eqnos` etc.). Since I suppose the main use case is using `pandoc-xnos` directly as a filter (only in cases where you'd specifically want to use *only* e.g. `eqnos` *and not* the others you'd use a single filter from the suite), the predominant use would probably still be as a standalone filter.

**Comment by ppenguin** [comment]
Ok, that should be it: `pandoc-{eq,seq,fig,table}nos` moved to `tools/misc` as `pythonApplication`, `pandoc-xnos` remains as `python3Package`.

File: pkgs/development/python-modules/pandoc-xnos/default.nix [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  };
+
+  propagatedBuildInputs = [ pandocfilters psutil ];
+
+  doCheck = false;
+
+  pythonImportsCheck = [ "pandocxnos" ];
+
+  meta = with lib; {
+    description = "The pandoc-xnos filter suite provides facilities for cross-referencing in markdown documents processed by pandoc.";
```
Comment (loicreynier) [inline pkgs/development/python-modules/pandoc-xnos/default.nix:27]:
```suggestion
    description = "Pandoc filter suite providing facilities for cross-referencing in markdown documents";
```

File: pkgs/tools/misc/pandoc-eqnos/default.nix [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    sha256 = "sha256-7GQdfGHhtQs6LZK+ZyMmcPSkoFfBWmATTMejMiFcS7Y=";
+  };
+
+  propagatedBuildInputs = [ pandoc-xnos ];
+
+  # Different pandoc executables are not available
+  doCheck = false;
+
+  meta = with lib; {
+    description = "Pandoc filter for numbering equations and their references";
```
Comment (loicreynier) [inline pkgs/tools/misc/pandoc-eqnos/default.nix:25]:
```suggestion
    description = "Standalone Pandoc filter from the pandoc-xnos suite for numbering equations and equation references";
```

File: pkgs/tools/misc/pandoc-fignos/default.nix [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    sha256 = "sha256-eDwAW0nLB4YqrWT3Ajt9bmX1A43wl+tOPm2St5VpCLk=";
+  };
+
+  propagatedBuildInputs = [ pandoc-xnos ];
+
+  # Different pandoc executables are not available
+  doCheck = false;
+
+  meta = with lib; {
+    description = "Pandoc filter for numbering figures and their references.";
```
Comment (loicreynier) [inline pkgs/tools/misc/pandoc-fignos/default.nix:25]:
```suggestion
    description = "Standalone Pandoc filter from the pandoc-xnos suite for numbering figures and figure references";
```

File: pkgs/tools/misc/pandoc-secnos/default.nix [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    sha256 = "sha256-5VgEq5NeKRexPLJyQ8Uwj3vfnN9LUGoVIUUMiaWzyxw=";
+  };
+
+  propagatedBuildInputs = [ pandoc-xnos ];
+
+  # Different pandoc executables are not available
+  doCheck = false;
+
+  meta = with lib; {
+    description = "Pandoc filter for numbering sections and their references.";
```
Comment (loicreynier) [inline pkgs/tools/misc/pandoc-secnos/default.nix:28]:
```suggestion    
    description = "Standalone Pandoc filter from the pandoc-xnos suite for numbering section references";
```

File: pkgs/tools/misc/pandoc-tablenos/default.nix [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    sha256 = "sha256-FwzsRziY3PoySo9hIFuLw6tOO9oQij6oQEyoY8HgnII=";
+  };
+
+  propagatedBuildInputs = [ pandoc-xnos ];
+
+  # Different pandoc executables are not available
+  doCheck = false;
+
+  meta = with lib; {
+    description = "Pandoc filter for numbering tables and their references.";
```
Comment (loicreynier) [inline pkgs/tools/misc/pandoc-tablenos/default.nix:25]:
```suggestion
    description = "Standalone Pandoc filter from the pandoc-xnos suite for numbering tables and table references";
```

**Review by loicreynier — COMMENTED** [review]
I left some `meta.description` suggestions to fit the [guidelines](https://nixos.org/manual/nixpkgs/stable/#var-meta-description) and to emphasis that `pandoc-`(`eqnos`, `fignos`, `secnos` and `tablenos`) are standalone filters.

**Comment by ppenguin** [comment]
@fabaff sorry for prodding, you all must be busy with 22.05 too... Is there anything remaining for me to do or is this one merge-ready now? (there indeed have been some more changes since last review, but all addressed issues and some others were addressed in the meantime...)

Cheers, ppenguin

**Review by SuperSandro2000 — COMMENTED** [review]
Please squash the commits together to one per package and one for the maintainer addition.

File: pkgs/tools/misc/pandoc-tablenos/default.nix [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    owner = "tomduck";
+    repo = pname;
+    rev = version;
+    sha256 = "sha256-FwzsRziY3PoySo9hIFuLw6tOO9oQij6oQEyoY8HgnII=";
+  };
+
+  propagatedBuildInputs = [ pandoc-xnos ];
+
+  # Different pandoc executables are not available
+  doCheck = false;
```
[the remark this answers is not in the document]
Comment (ppenguin) [reply pkgs/tools/misc/pandoc-tablenos/default.nix:22]: If I understand correctly these are simply stubs calling `xnos`, so the reason to keep them separate upstream (pip) would merely be to select fine-grained capability, even though from a resource-usage perspective it hardly makes a difference.
But I guess it's cleanest to just mirror the package units/scope from upstream.

File: pkgs/tools/misc/pandoc-secnos/default.nix [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  format = "pyproject";
+
+  src = fetchgit {
+    # FIXME: after PR to fix manifest accepted upstream, go back to upstream
+    # owner = "tomduck";
+    # repo = pname;
+    url = "https://github.com/ppenguin/pandoc-secnos.git";
+    rev = "165ee1f4c1208636254392335d34934dc50d273e";
+    sha256 = "sha256-5VgEq5NeKRexPLJyQ8Uwj3vfnN9LUGoVIUUMiaWzyxw=";
+  };
```
[the remark this answers is not in the document]
Comment (ppenguin) [reply pkgs/tools/misc/pandoc-secnos/default.nix]: done

File: pkgs/development/python-modules/pandoc-xnos/default.nix [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    owner = "tomduck";
+    repo = pname;
+    rev = version;
+    sha256 = "sha256-beiGvN0DS6s8wFjcDKozDuwAM2OApX3lTRaUDRUqLeU=";
+  };
+
+  propagatedBuildInputs = [ pandocfilters psutil ];
+
+  doCheck = true;
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pandoc-xnos/default.nix:23]:
```suggestion
```
We don't need to repeat the default.

**Pull request merged**